examples/getting_started/tools_and_agents.ipynb

```diff
@@ -0,0 +1,583 @@
+    "\n",
+    "For more details on implementing and using Tools in Mirascope, see the [Tools documentation](https://docs.mirascope.io/latest/learn/tools/)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 5. Creating Custom Tools\n",
+    "\n",
+    "Now, let's create a more complex Tool that performs web searches using the DuckDuckGo search engine. We'll implement this tool as a method within our agent class to demonstrate how tools can access the agent's state:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2024-09-11T08:22:28.087233Z",
+     "start_time": "2024-09-11T08:22:28.030706Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "from duckduckgo_search import DDGS\n",
+    "\n",
+    "\n",
+    "class WebSearchAgentBase(BaseModel):\n",
+    "    messages: list[BaseMessageParam | openai.OpenAIMessageParam] = []\n",
+    "    search_history: list[str] = []\n",
+    "    max_results_per_query: int = 2\n",
+    "\n",
+    "    def web_search(self, queries: list[str]) -> str:\n",
+    "        \"\"\"Performs web searches for given queries and returns URLs.\n",
+    "\n",
+    "        Args:\n",
+    "            queries: List of search queries.\n",
+    "\n",
+    "        Returns:\n",
+    "            str: Newline-separated URLs from search results or error messages.\n",
+    "\n",
+    "        Raises:\n",
+    "            Exception: If web search fails entirely.\n",
+    "        \"\"\"\n",
+    "        try:\n",
+    "            urls = []\n",
+    "            for query in queries:\n",
+    "                results = DDGS(proxies=None).text(\n",
+    "                    query, max_results=self.max_results_per_query\n",
+    "                )\n",
+    "                for result in results:\n",
+    "                    link = result[\"href\"]\n",
+    "                    try:\n",
+    "                        urls.append(link)\n",
+    "                    except Exception as e:\n",
+    "                        urls.append(\n",
+    "                            f\"{type(e)}: Failed to parse content from URL {link}\"\n",
+    "                        )\n",
+    "            self.search_history.extend(queries)\n",
+    "            return \"\\n\\n\".join(urls)\n",
+    "        except Exception as e:\n",
+    "            return f\"{type(e)}: Failed to search the web for text\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This `web_search` method is a more complex Tool that takes a list of search queries and returns a string of newline-separated URLs from the search results. It uses the DuckDuckGo search API to perform the searches and updates the agent's `search_history`.\n",
```

Comment:
nit: add something about "Luckily the DuckDuckGO search API does not require an API key!"